app/controllers/locations_controller.rb

```diff
@@ -0,0 +1,57 @@
+class LocationsController < ApplicationController
+  before_action :authenticate_user!
+  before_action :authenticate_scope!
+  before_action :find_location, except: %i[new create]
+  before_action :find_scheme
+  before_action :authenticate_action!
+
+  def new
+    @location = Location.new
+  end
+
+  def create
+    @location = Location.new(location_params)
+
+    if @location.save
+      location_params[:add_another_location] == "Yes" ? redirect_to(location_new_scheme_path) : redirect_to(scheme_check_answers_path(scheme_id: @scheme.id))
+    else
+      render :new, status: :unprocessable_entity
+    end
+  end
+
+  def details; end
+
+  def update
+    if @location.update(location_params)
+      location_params[:add_another_location] == "Yes" ? redirect_to(location_new_scheme_path) : redirect_to(scheme_check_answers_path(@scheme, anchor: "locations"))
+    else
+      render :details, status: :unprocessable_entity
+    end
+  end
+
+private
+
+  def find_scheme
+    @scheme = params[:scheme_id] && params[:id] ? Scheme.find(params[:scheme_id]) : Scheme.find(params[:id])
```

Comment:
Should `params[:id]` ever refer to scheme on a request to the `locations` controller? That seems like it's wrong?

Reply:
Good point, will review my routing.